Cgranade/0.6 (#408)
MicrosoftDocs/quantum-docs-pr: articles/libraries/standard/diagnostics.md

---
title: Q# standard libraries - diagnostics | Microsoft Docs
description: Q# standard libraries - diagnostics
author: cgranade
uid: microsoft.quantum.libraries.diagnostics
ms.author: chgranad@microsoft.com 
ms.topic: article
---

# Diagnostics #

As with classical development, it is important to be able to diagnose mistakes and errors in quantum programs.
The Q# standard libraries provide a variety of different ways to ensure the correctness of quantum programs, as detailed in <xref:microsoft.quantum.techniques.testing-and-debugging>.

> Largely speaking, this support comes in the form of functions and operations that either instruct target machines to provide additional diagnostic information to host programs, or that enforce the correctness of conditions and invariants expressed by quantum programmers.

Reviewer: ```suggestion
Largely speaking, this support comes in the form of functions and operations that either instruct the target machine to provide additional diagnostic information to the host program or developer, or enforce the correctness of conditions and invariants expressed by the function or operation call.
```

## Machine Diagnostics ##

Diagnostics about classical values can be obtained by using the <xref:microsoft.quantum.intrinsic.message> function to log a message in a machine-dependent way.
By default, this writes the string to the console.
Used together with interpolated strings, <xref:microsoft.quantum.intrinsic.message> makes it easy to report diagnostic information about classical values:

```Q#
let angle = Microsoft.Quantum.Math.PI() * 2.0 / 3.0;
Message($"About to rotate by an angle of {angle}...");
```

> [!NOTE]

> > `Message` has signature `((String) -> Unit)`, again representing that emitting a debug log message cannot be observed from within Q#.

Reviewer: ```suggestion
> `Message` has signature `(String -> Unit)`, again representing that emitting a debug log message cannot be observed from within Q#.
```

> The <xref:microsoft.quantum.diagnostics.dumpmachine> and <xref:microsoft.quantum.diagnostics.dumpregister> callables instruct target machines to provide diagnostic information about all currently allocated qubits or specific register of qubits, respectively.

Reviewer: ```suggestion
The <xref:microsoft.quantum.diagnostics.dumpmachine> and <xref:microsoft.quantum.diagnostics.dumpregister> callables instruct target machines to provide diagnostic information about all currently allocated qubits or about a specific register of qubits, respectively.
```
